walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Arrays;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<Event> eventsAL = new ArrayList<Event>(events);
+      Collection<String> people = request.getAttendees();
+
+      Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+      availableTimes.add(TimeRange.WHOLE_DAY);
+      getAvailableTimes(eventsAL, people, availableTimes, request.getDuration());
+    
+      Collection<TimeRange> availableTimesOptional = new ArrayList<TimeRange>(availableTimes);
+      Collection<String> optionalPeople = request.getOptionalAttendees();
+      getAvailableTimes(eventsAL, optionalPeople, availableTimesOptional, request.getDuration());
+
+      if(request.getAttendees().isEmpty() && availableTimesOptional.isEmpty()){
+          return Arrays.asList();
+      }
+
+      return availableTimesOptional.isEmpty() ? availableTimes : availableTimesOptional;
+  }
+
+  private void getAvailableTimes(Collection<Event> events, Collection<String> people, Collection<TimeRange> availableTimes, long duration){
+      Iterator<Event> it = events.iterator();
+      while(it.hasNext()){
+          Event element = it.next();
+          if(!Collections.disjoint(element.getAttendees(), people)){
+            createAvailableRange(availableTimes, element.getWhen());
+            it.remove();
+          }
+      }
+      availableTimes.removeIf(item -> (item.duration() < duration));
```

Comment:
We can filter the duration in the same pass (we don't need to do another pass for this).

If you're interested in the java 8 stream API, then that may also come in handy as well.

Reply:
I highly recommend streams... lifechanging :)

Reply:
done in one pass now